src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -148,14 +158,19 @@ public void CompleteStream(bool errored)
                     }
                 }
 
-                _http2Output.Dispose();
+                _http2Output.Complete();
 
                 RequestBodyPipe.Reader.Complete();
 
                 // The app can no longer read any more of the request body, so return any bytes that weren't read to the
                 // connection's flow-control window.
                 _inputFlowControl.Abort();
 
+                // We only want to reuse a stream that has completely finished writing.
```

Comment:
What's the deal with this?

Reply:
@halter73 wanted to check these values before reusing a stream. HasResponseCompleted ensures the output producer task has been awaited

Reply:
Why would we return the stream before then?

Reply:
If an app didn't await a write, and Kestrel didn't try to flush a suffix.

Not very likely, but you could imagine some app doing some custom Task.WhenAny/Task.Delay logic on a response body write, and then throwing if the delay finishes before the write. Of course the app *should* just use a cancellation token, but I don't see the harm in guarding against this.